homeassistant/components/zwave_js/discovery.py

```diff
@@ -0,0 +1,130 @@
+"""Map Z-Wave nodes and values to Home Assistant entities."""
+
+from dataclasses import dataclass, field
+from typing import Any, AsyncGenerator, List, Optional
+
+from zwave_js_server.const import CommandClass
+from zwave_js_server.model.node import Node as ZwaveNode
+from zwave_js_server.model.value import Value as ZwaveValue
+
+
+@dataclass
+class ZwaveDiscoveryInfo:
+    """Info discovered from (primary) ZWave Value to create entity."""
+
+    node: ZwaveNode  # node to which the value(s) belongs
+    primary_value: ZwaveValue  # the value object itself for primary value
+    platform: str  # the home assistant platform for which an entity should be created
+    platform_hint: str = ""  # hint for the platform about this discovered entity
+
+    @property
+    def value_id(self):
+        """Return the unique value_id belonging to primary value."""
+        return f"{self.node.node_id}.{self.primary_value.value_id}"
+
+
+@dataclass
+class ZWaveDiscoverySchema:
+    """Z-Wave discovery schema.
+
+    The (primary) value for an entity must match these conditions.
+    """
+
+    # specify the hass platform for which this scheme applies (e.g. light, sensor)
+    platform: str
+    # [optional] hint for platform
+    hint: Optional[str]
+    # [optional] the node's basic device class must match ANY of these values
+    device_class_basic: List[str] = field(default_factory=list)
+    # [optional] the node's generic device class must match ANY of these values
+    device_class_generic: List[str] = field(default_factory=list)
+    # [optional] the node's specific device class must match ANY of these values
+    device_class_specific: List[str] = field(default_factory=list)
+    # [optional] the value's command class must match ANY of these values
+    command_class: List[int] = field(default_factory=list)
+    # [optional] the value's endpoint must match ANY of these values
+    endpoint: List[int] = field(default_factory=list)
+    # [optional] the value's property must match ANY of these values
+    property: List[str] = field(default_factory=list)
+    # [optional] the value's metadata_type must match ANY of these values
+    type: List[str] = field(default_factory=list)
```

Comment:
All these defaults will generate a lot of empty lists. I would prefer we just do a `None` check. Also would a `set` be better? 

```python
disc.endpoint is None or value.endpoint in disc.endpoint
```

Reply:
This is easy to read for new contributors that want to add a device implementation to the discovery schema. The previous scheme was scary ;-)

I'd say we adjust it to a set instead.

Reply:
we could also set each attribute to None by default and the type will be Optional[List[str]]
that saves having all these empty lists

Reply:
@MartinHjelmare what do to with this ? use a set as default initializer or None value ?

Reply:
Can we do Optional set and default to None?

Reply:
I'll do Optional List because that is way easier to initialize in the discovery scheme, ok ?

Reply:
![Schermafbeelding 2021-01-08 om 14 26 17](https://user-images.githubusercontent.com/6389780/104020411-79427780-51bd-11eb-8912-7ff04d2a80e7.png)

Reply:
done